contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -0,0 +1,390 @@
+        address paymasterId,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint32 priceMarkup
+    ) public view returns (bytes32) {
+        //can't use userOp.hash(), since it contains also the paymasterAndData itself.
+        return
+            keccak256(
+                abi.encode(
+                    userOp.getSender(),
+                    userOp.nonce,
+                    userOp.initCode,
+                    userOp.callData,
+                    userOp.callGasLimit,
+                    userOp.verificationGasLimit,
+                    userOp.preVerificationGas,
+                    userOp.maxFeePerGas,
+                    userOp.maxPriorityFeePerGas,
+                    block.chainid,
+                    address(this),
+                    paymasterId,
+                    validUntil,
+                    validAfter,
+                    priceMarkup
+                )
+            );
+    }
+
+    // Note: do not use this in validation phase
+    function getGasPrice(
+        uint256 maxFeePerGas,
+        uint256 maxPriorityFeePerGas
+    ) internal view returns (uint256) {
+        if (maxFeePerGas == maxPriorityFeePerGas) {
+            //legacy mode (for networks that don't support basefee opcode)
+            return maxFeePerGas;
+        }
+        return minuint256(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
+    }
+
+    function minuint256(
+        uint256 a,
+        uint256 b
+    ) internal pure returns (uint256 result) {
+        assembly {
+            result := xor(b, mul(xor(b, a), gt(a, b)))
+        }
+    }
+
+    function maxuint256(
+        uint256 a,
+        uint256 b
+    ) internal pure returns (uint256 result) {
+        assembly {
+            result := xor(a, mul(xor(a, b), gt(b, a)))
+        }
+    }
+
+    function minuint32(
+        uint32 a,
+        uint32 b
+    ) internal pure returns (uint32 result) {
+        assembly {
+            result := xor(b, mul(xor(b, a), gt(a, b)))
+        }
+    }
+
+    function maxuint32(
+        uint32 a,
+        uint32 b
+    ) internal pure returns (uint32 result) {
+        assembly {
+            result := xor(a, mul(xor(a, b), gt(b, a)))
+        }
+    }
+
+    /**
+     * @dev Verify that an external signer signed the paymaster data of a user operation.
+     * The paymaster data is expected to be the paymaster and a signature over the entire request parameters.
+     * @param userOp The UserOperation struct that represents the current user operation.
+     * userOpHash The hash of the UserOperation struct.
+     * @param requiredPreFund The required amount of pre-funding for the paymaster.
+     * @return context A context string returned by the entry point after successful validation.
+     * @return validationData An integer returned by the entry point after successful validation.
+     */
+    function _validatePaymasterUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash,
```

Comment:
userOpHash is not being used here

Reply:
I recently removed it from event and context. will omit here also again